grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy

```diff
@@ -272,34 +273,50 @@ class ProductSupplierService {
 
         PreferenceType preferenceType = params.globalPreferenceTypeName ? PreferenceType.findByName(params.globalPreferenceTypeName) : null
 
-        if (preferenceType) {
-            ProductSupplierPreference productSupplierPreference = productSupplier.getGlobalProductSupplierPreference()
+        assignDefaultPreferenceType(productSupplier,
+                preferenceType,
+                params.globalPreferenceTypeComments,
+                params.globalPreferenceTypeValidityStartDate,
+                params.globalPreferenceTypeValidityEndDate)
 
-            if (!productSupplierPreference) {
-                productSupplierPreference = new ProductSupplierPreference()
-                productSupplier.addToProductSupplierPreferences(productSupplierPreference)
-            }
+        if (!productSupplier.code && !productSupplier.id) {
+            assignSourceCode(productSupplier, supplier)
+        }
+        return productSupplier
+    }
 
-            productSupplierPreference.preferenceType = preferenceType
-            productSupplierPreference.comments = params.globalPreferenceTypeComments
+    void assignDefaultPreferenceType(ProductSupplier productSupplier,
+                 PreferenceType preferenceType,
+                 String comments,
+                 String validityStartDate,
+                 String validityEndDate) {
+        ProductSupplierPreference productSupplierPreference = productSupplier.getGlobalProductSupplierPreference()
+        if (!preferenceType) {
+            // If preference type is not provided, delete it
+            productSupplier.removeFromProductSupplierPreferences(productSupplierPreference)
+            productSupplierPreference.delete()
+            return
+        }
 
-            def globalPreferenceTypeValidityStartDate = params.globalPreferenceTypeValidityStartDate ? dateFormat.parse(params.globalPreferenceTypeValidityStartDate) : null
+        if (!productSupplierPreference) {
+            productSupplierPreference = new ProductSupplierPreference()
+            productSupplier.addToProductSupplierPreferences(productSupplierPreference)
+        }
 
-            if (globalPreferenceTypeValidityStartDate) {
-                productSupplierPreference.validityStartDate = globalPreferenceTypeValidityStartDate
-            }
+        productSupplierPreference.preferenceType = preferenceType
+        productSupplierPreference.comments = comments
 
-            def globalPreferenceTypeValidityEndDate = params.globalPreferenceTypeValidityEndDate ? dateFormat.parse(params.globalPreferenceTypeValidityEndDate) : null
+        Date globalPreferenceTypeValidityStartDate = validityStartDate ? Constants.EXPIRATION_DATE_FORMATTER.parse(validityStartDate) : null
```

Comment:
I was supposed to use an existing constant. I would say technically it might be considered as the expiration date, as the validity date is somewhat a synonym, isn't it? :thinking: